text/trace/0205-messaging-semantic-conventions-context-propagation.md

```diff
@@ -0,0 +1,173 @@
+message.
+
+Although OpenTelemetry semantic conventions cannot specify the exact mechanisms
+to achieve this for every intermediary and every protocol, this requirement
+must be clearly formulated, so that it can be implemented by protocols and
+instrumentations.
+
+## Proposed addition to the messaging semantic conventions
+
+### Context propagation
+
+A message may pass many different components and layers in one or more
+intermediaries when it is propagated from the producer to the consumer. It
+cannot be assumed, and in many cases, it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _creation context_ allows correlating the producer with the consumers of a
+message, regardless of intermediary instrumentation. The creation context is
+created by the producer and must be propagated to the consumers. It must not be
+altered by intermediaries.  This context helps to model dependencies between
+producers and consumers, regardless of the underlying messaging transport
+mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extracted from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+### Requirements
+
+A producer SHOULD attach a creation context to each message. The creation context
+SHOULD be attached in a way so that it is not possible to be changed by intermediaries.
+
+## Future possibilities
+
+### Transport context propagation
+
+As illustrated in the [example](#example) above, a creation context can be
+attached to a message, while different contexts are propagated via HTTP requests
+publishing and fetching the message. When coming up with conventions and
+guidance for intermediary instrumentation, it will be beneficial to clearly
+outline those two layers of context propagation and build conventions for
+intermediary instrumentation on top of this outline:
+
+1. The _creation context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. An additional _transport context layer_ allows correlating the producer and
+   the consumer with an intermediary. It also allows correlating multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries that
+   are not instrumented might simply drop the transport context.
```

Comment:
I know this is a "future", but I think intermediaries that are not instrumented might pass through the transport context unaltered. 

If we consider a message following a path:
producer --> messaging SDK --> intermediary --> messaging SDK --> consumer

If the SDK on the producer side includes a transport context and the intermediary isn't instrumented, having the transport context available to the consumer SDK could be useful. The SDK may instrument details like "received from intermediary", "enqueued to application", "delivered to application". I think these should be considered part of the transport trace, even if the intermediary didn't generate spans and modify the context.

Reply:
Not sure if I follow but if the intermediary is not instrumented, how would the transport trace continue? Somehow there must be some code that reads the incoming remote span and either creates a new child span or simply forwarded the same (which I'm not it's a good approach).

Reply:
I think the main point we wanted to make here was, that we cannot make any strong assumptions regarding the transport context. By definition, intermediaries can change it. So, possibly, intermediaries might alter it, keep it as it is (that's the scenario @dpauls points out), or drop it. The main point we wanted to make is, that in contrast to the message context, the consumer cannot rely on receiving a transport context, even if the producer sent one.

I can also remove the sentence in question. I think at this point we don't want to make any assumptions or requirements about what un-instrumented intermediaries would or should do.

Reply:
> Not sure if I follow but if the intermediary is not instrumented, how would the transport trace continue? 
If an intermediary doesn't even understand the concept of tracing, if it sees a message with a `traceparent` header or property, it is probably going to treat it the same as it would a `foo` or a `bar` header/property. Different intermediaries might do different things here, but a reasonable choice is to just propagate those through unchanged towards the consumer.

I agree with all three possibilities (alter, propagate, drop) that @pyohannes mentioned. While this last sentence uses a "might" which is in no way prescriptive, I think it makes sense to either mention all possibilities, or also removing the sentence would be reasonable as well. Singling out the "drop" scenario might cause some readers to interpret this as being a preferable behaviour.

Reply:
I removed the last sentence of this paragraph.